db-tests/src/test/java/gov/nasa/jpl/aerie/database/ExternalEventTests.java

```diff
@@ -0,0 +1,1922 @@
+      }
+    }
+  }
+  //endregion
+
+
+  //region Derivation Tests
+  /**
+   * The focus of this class is to test specifically the derivation function for events. This includes overlapping of
+   *    source windows, as well as reconciling differences inside the sources themselves (by testing each of the 4
+   *    derivation rules).
+   */
+  @Nested
+  @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+  class DerivedEventsTests {
+
+    /**
+     * This class focuses on testing sources' windows to verify that derivation will work as expected. These are
+     *    separate, extracted tests that don't really evaluate events and instead just that sources play together
+     *    correctly. As such we test "empty" sources to make sure their overlapped windows work correctly.
+     * We add events that span a very short duration simply so that the sources show up in derived_events, but we aren't
+     *    testing any properties of said events.
+     */
+    @Nested
+    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+    class DerivedSourcesTests {
+
+      // Commonly Repeated:
+      String duration = "00:00:00.000001";
+
+      /**
+       * The first test is a basic, non-overlapping case. We must ensure that gaps are preserved:
+       *
+       *    A:             ++++++++
+       *    B:  +++++++
+       *        BBBBBBB    AAAAAAAA
+       */
+      @Test
+      void testSparseCoverage() throws SQLException {
+        try(final var statement = connection.createStatement()) {
+
+          // create our sources and their per-window events
+          ExternalSource a = new ExternalSource(
+              "A",
+              st,
+              dg,
+              "2024-01-01T00:00:00Z",
+              "2024-01-01T03:00:00Z",
+              "2024-01-01T04:00:00Z",
+              ca,
+              mt
+          );
+          ExternalEvent aE = new ExternalEvent(a.key() + "_event", st, a.key(), dg, a.start_time(), duration, mt);
+          ExternalSource b = new ExternalSource(
+              "B",
+              st,
+              dg,
+              "2024-01-02T00:00:00Z",
+              "2024-01-01T01:00:00Z",
+              "2024-01-01T02:00:00Z",
+              ca,
+              mt
+          );
+          ExternalEvent bE = new ExternalEvent(b.key() + "_event", st, b.key(), dg, b.start_time(), duration, mt);
+
+          // verify the ranges are as expected
+          assertDoesNotThrow(() -> {
+            // insert generic external event type, source type, and derivation group
+            insertStandardTypes();
+
+            // insert sources
+            insertExternalSource(a);
+            insertExternalSource(b);
+
+            // insert events
+            insertExternalEvent(aE);
+            insertExternalEvent(bE);
+
+            var res = statement.executeQuery(
+                // language=sql
+                """
+                SELECT * FROM merlin.derived_events ORDER BY source_key
+                """
+            );
```

Comment:
Extract this repeated code into a method `getDerivedEvents` that returns a List of `DerivedEvents`. Then have `ExpectedResults` objects be lists of these events. Then compare these two lists instead of using your `compareLists` method (or update `compareLists` to actually take two lists)

Reply:
Would like to add that using a proper `DerivedEvents` record that includes more than just one column would make the test stronger.

Reply:
Resolved.

Reply:
Unresolved, you did not fully address my comment.

Reply:
Calls to `compareLists` have been replaced with `getDerivedEvents()` and the recommended `assertTrue(actualResults.containsAll(List.of(expectedResults)));`. 

The only unresolved part is not making `expectedResults` also a list of these events, but that introduces a lot of unnecessary bloat; unnecessarily creating entire event objects when the only object of comparison is ranges seems unnecessary.

Reply:
The bloat is not unnecessary as it makes it very explicit what the expected row from the database is, making the test stronger vs just checking one column (current behavior)

Reply:
This allows you to definitively state that each range corresponds to the expected source, instead of assuming as much based on the ranges roughly lining up.

Reply:
Fixed in [expand derivedEvents tests to use DerivedEvent objects](https://github.com/NASA-AMMOS/aerie/commit/bdfda028fa513fd3c82d177f5f49248458c40788).